// @inversifyjs/core/lib/esm/metadata/calculations/getClassMetadataFromMetadataReader.js
import { getReflectMetadata } from '@inversifyjs/reflect-metadata-utils';
import { POST_CONSTRUCT, PRE_DESTROY } from '../../reflectMetadata/data/keys';
import { getClassMetadataConstructorArgumentsFromMetadataReader } from './getClassMetadataConstructorArgumentsFromMetadataReader';
import { getClassMetadataPropertiesFromMetadataReader } from './getClassMetadataPropertiesFromMetadataReader';
export function getClassMetadataFromMetadataReader(type, metadataReader) {
    const postConstructMetadata = getReflectMetadata(type, POST_CONSTRUCT);
    const preDestroyMetadata = getReflectMetadata(type, PRE_DESTROY);
    const classMetadata = {
        constructorArguments: getClassMetadataConstructorArgumentsFromMetadataReader(type, metadataReader),
        lifecycle: {
            postConstructMethodName: postConstructMetadata?.value,
            preDestroyMethodName: preDestroyMetadata?.value,
        },
        properties: getClassMetadataPropertiesFromMetadataReader(type, metadataReader),
    };
    return classMetadata;
}
